laborator/lab-02/README.md

```diff
@@ -0,0 +1,554 @@
+     swap(&a, &b);
+ 
+     printf("a = %d, b = %d\n", a, b); /* Se va afișa a = 5, b = 3 */
+ 
+     return 0;
+}
+```
+
+O funcție poate returna un pointer, dar acest pointer nu poate conține adresa unei variabile locale. De cele mai multe ori, rezultatul este unul din argumente, modificat eventual în funcție. Spre exemplu:
+
+```c
+char* toUpper(char *s) {
+    /* Primește un sir de caractere și întoarce șirul scris cu majuscule */
+    for (int i = 0 ; s[i] ; ++i) {
+        if (s[i] >= 'a' && s[i] <= 'z') {
+            s[i] -= 32;
+        }
+    }
+ 
+    return s;
+}
+```
+
+Dacă o funcție returnează adresa unei variabile locale este obligatoriu ca aceasta să fie statică. Durata de viață a unei variabile locale se încheie odată cu terminarea execuției funcției în care a fost definită și de aceea adresa unei astfel de variabile nu trebuie transmisă în afara funcției.
+
+Numele unei funcții reprezintă adresa de memorie la care începe funcția. Un pointer la o funcție este o variabilă ce stochează adresa unei funcții ce poate fi apelată ulterior prin intermediul acelui pointer. Uzual, pointerii la funcții sunt folosiți pentru a trimite o funcție ca parametru unei alte funcții.
+
+Declararea unui pointer la o funcție se face în felul următor: **tip (\*pf) (lista_parametri_formali)**
+
+De ce este necesară folosirea parantezelor suplimentare? Dacă acestea ar lipsi atunci am discuta despre o funcție ce are ca rezultat un pointer. În continuare, sunt prezentate doua exemple de folosire a pointerilor la funcții:
+
+```c
+int add(int a, int b) {
+    return a + b;
+}
+ 
+int subtract(int a, int b) {
+    return a - b;
+}
+ 
+int operation(int x, int y, int (*func) (int, int)) {
+    return func(x, y);
+}
+ 
+int main() {
+    int (*minus)(int, int) = subtract;
+    printf("%d", operation(10, 5, minus)); /* Se va afișa 5 */
+ 
+    return 0;
+}
+```
+
+Funcția [qsort()](http://www.cplusplus.com/reference/cstdlib/qsort/) din `stdlib.h` folosește drept comparator un pointer la funcție.
+
+```c
+int compare(const void *a, const void *b) {
+    return *(int *) a - *(int *)b;
+}
+ 
+int main() {
+    int v[] = {100, 5, 325, 1, 30};
+    int size = sizeof(v) / sizeof(*v);
+ 
+    qsort(v, size, sizeof(*v), compare);
+    for (int i = 0 ; i < size ; ++i) {
+        printf("%d ", v[i]);
+    }
+ 
+    return 0;
+}
+```
+
+## **GNU Debugger(GDB)**
+
+**Lansarea în execuție a programului**
+
+Pentru a lansa programul urmărit în execuție există două comenzi disponibile:
+
+- **run** - această comandă va lansa în execuție programul
+- **start** - spre deosebire de run, această comandă va începe execuția programului, însă se va opri imediat după intrarea în main
+
+**Breakpoints**
+
+Elementul esențial al GDB-ului este breakpoint-ul. Practic, un breakpoint setat la o anumită instrucțiune face ca execuția programului să se oprească de fiecare dată când se ajunge la acest punct. Setarea unui breakpoint se face cu următoarea comandă:
+
+```py
```

Comment:
Markdown-ul de pe GitHub folosește linguist pentru syntax highlighting (conform paginii de [aici](https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks)). La finalul paginii este menționat un fișier yaml cu limbajele suportate, printre care este enumerat și [gdb](https://github.com/github/linguist/blob/master/lib/linguist/languages.yml#L1854). Încearcă să vezi cum arată.